app/code/Magento/Eav/Model/Entity/AbstractEntity.php

```diff
@@ -1958,4 +1959,25 @@ protected function loadAttributesForObject($attributes, $object = null)
             }
         }
     }
+
+    /**
+     * Receive a list of EAV attributes using provided metadata service.
+     *
+     * @param MetadataServiceInterface $metadataService
+     * @return string[]
+     */
+    protected function getEavAttributesCodes(MetadataServiceInterface $metadataService)
```

Comment:
Inheritance based API not encouraged in Magento 2

Reply:
I know but I have to work with what is there (which is inheritance based resourece models) or introduce huge changes, probably BIC. Do you have an idea for a simple solution?

Reply:
it's basicalls moved from `\Magento\Framework\Model\AbstractExtensibleModel::getEavAttributesCodes` (inheritance based model API)

A possible non-inheritance solution I see is to to implement the method in the Category and Product resource models seperately in a non-generic way. Then, if we want to get rid of duplicated logic, add a method `getCustomAttributeCodes($dataObjectClassName)` to `\Magento\Framework\Api\MetadataServiceInterface` (BIC: new method in interface)

Reply:
disregard this, I've read your top comment